torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1443,55 +1402,39 @@ void InitXlaModuleBindings(py::module m) {
-          options.conv_halo_exchange_always_on_lhs =
-              conv_halo_exchange_always_on_lhs;
-          options.allow_module_signature_change = true;
-          options.choose_faster_windowed_einsum_over_mem =
-              choose_faster_windowed_einsum;
-          options.unroll_windowed_einsum = unroll_windowed_einsum;
-          options.bidirectional_windowed_einsum = bidirectional_windowed_einsum;
-
-          xla::HloModuleConfig config;
-          config.set_use_spmd_partitioning(true);
-          config.set_replica_count(num_replicas);
-          config.set_num_partitions(num_devices);
-
-          auto hlo_text = GetTensorsHloGraph(tensors);
-          auto hlo_module_error =
-              xla::ParseAndReturnUnverifiedModule(hlo_text, config);
-          if (!hlo_module_error.ok()) {
-            LOG(ERROR) << "HLO Module loading failed: "
-                       << hlo_module_error.status();
-            return nullptr;
-          }
-          auto module = std::move(hlo_module_error.ValueOrDie());
-
-          auto collective_ops_creator =
-              xla::spmd::GetDefaultCollectiveOpsCreator(
-                  num_devices, /*num_replicas=*/num_replicas);
-
-          xla::HloPassPipeline pass("spmd-partitioning");
-          pass.AddPass<xla::HloVerifier>(/*layout_sensitive=*/false,
-                                         /*allow_mixed_precision=*/false);
-          pass.AddPass<xla::ShardingPropagation>(/*is_spmd=*/true);
-          pass.AddPass<xla::spmd::SpmdPartitioner>(
-              num_devices, /*num_replicas=*/num_replicas, options,
-              collective_ops_creator);
-          pass.AddPass<xla::HloVerifier>(/*layout_sensitive=*/false,
-                                         /*allow_mixed_precision=*/false);
-          const auto& pass_status = pass.Run(module.get());
-          if (!pass_status.ok()) {
-            XLA_ERROR() << "spmd-partitioning pass failed";
-          }
-          return module->ToString();
-        });
+  // This is useful for debugging and generating a partitioned HLO separately
+  // outside the actual compilation & execution. This allows testing with
+  // different partitioning configurations.
+  m.def(
+      "_xla_partitioning_pass",
+      [](const std::vector<at::Tensor>& tensors, int64_t num_replicas,
+         int64_t num_devices, bool conv_halo_exchange_always_on_lhs = true,
+         bool choose_faster_windowed_einsum = false,
+         bool unroll_windowed_einsum = false,
+         bool bidirectional_windowed_einsum = false) -> std::string {
+        xla::HloModuleConfig config;
+        config.set_use_spmd_partitioning(true);
+        config.set_replica_count(num_replicas);
+        config.set_num_partitions(num_devices);
+
+        std::string hlo_text = GetTensorsHloGraph(tensors);
+        auto hlo_module_error =
+            xla::ParseAndReturnUnverifiedModule(hlo_text, config);
+        if (!hlo_module_error.ok()) {
```

Comment:
should we just error out here with `XLA_CHECK_OK`?

Reply:
I like the idea, ack.

Reply:
Done